File: petties_mobile/lib/data/services/vet_shift_service.dart
import '../models/vet_shift.dart';
import 'api_client.dart';

class VetShiftService {
  final ApiClient _apiClient;

  VetShiftService({ApiClient? apiClient})
      : _apiClient = apiClient ?? ApiClient();

  /// Get shifts for the currently logged-in vet
  Future<List<VetShiftResponse>> getMyShifts({
    required String startDate,
    required String endDate,
  }) async {
    final response = await _apiClient.get(
      '/shifts/me',
      queryParameters: {
        'startDate': startDate,
        'endDate': endDate,
      },
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = response.data;
      return data.map((json) => VetShiftResponse.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load shifts');
    }
  }

  /// Get shift detail with slots
  Future<VetShiftResponse> getShiftDetail(String shiftId) async {
    final response = await _apiClient.get('/shifts/$shiftId');

    if (response.statusCode == 200) {
      return VetShiftResponse.fromJson(response.data);
    } else {
      throw Exception('Failed to load shift detail');
    }
  }

  /// Block a slot
  Future<SlotResponse> blockSlot(String slotId) async {
    final response = await _apiClient.patch('/slots/$slotId/block');

    if (response.statusCode == 200) {
      return SlotResponse.fromJson(response.data);
    } else {
      throw Exception('Failed to block slot');
    }
  }

  /// Unblock a slot
  Future<SlotResponse> unblockSlot(String slotId) async {
    final response = await _apiClient.patch('/slots/$slotId/unblock');

    if (response.statusCode == 200) {
      return SlotResponse.fromJson(response.data);
    } else {
      throw Exception('Failed to unblock slot');
    }
  }
}

File: petties-web/src/services/api/vetShiftService.ts
import { apiClient } from './client'
import type { VetShiftRequest, VetShiftResponse, SlotResponse } from '../../types/vetshift'

export const vetShiftService = {
    /**
     * Create one or more vet shifts
     */
    createShift: async (clinicId: string, data: VetShiftRequest): Promise<VetShiftResponse[]> => {
        const response = await apiClient.post<VetShiftResponse[]>(`/clinics/${clinicId}/shifts`, data)
        return response.data
    },

    /**
     * Get all shifts for a clinic in a date range
     */
    getShiftsByClinic: async (
        clinicId: string,
        startDate: string,
        endDate: string
    ): Promise<VetShiftResponse[]> => {
        const params = new URLSearchParams({
            startDate,
            endDate,
        })
        const url = `/clinics/${clinicId}/shifts?${params.toString()}`
        const response = await apiClient.get<VetShiftResponse[]>(url)
        return response.data
    },

    /**
     * Get shifts for the currently logged-in vet
     */
    getMyShifts: async (startDate: string, endDate: string): Promise<VetShiftResponse[]> => {
        const params = new URLSearchParams({ startDate, endDate })
        const response = await apiClient.get<VetShiftResponse[]>(`/shifts/me?${params.toString()}`)
        return response.data
    },

    /**
     * Get shift detail with slots
     */
    getShiftDetail: async (shiftId: string): Promise<VetShiftResponse> => {
        const response = await apiClient.get<VetShiftResponse>(`/shifts/${shiftId}`)
        return response.data
    },

    /**
     * Delete a shift
     */
    deleteShift: async (shiftId: string): Promise<void> => {
        await apiClient.delete(`/shifts/${shiftId}`)
    },

    /**
     * Delete multiple shifts
     */
    bulkDeleteShifts: async (shiftIds: string[]): Promise<void> => {
        await apiClient.delete('/shifts/bulk', { data: shiftIds })
    },

    /**
     * Block a slot
     */
    blockSlot: async (slotId: string): Promise<SlotResponse> => {
        const response = await apiClient.patch<SlotResponse>(`/slots/${slotId}/block`)
        return response.data
    },

    /**
     * Unblock a slot
     */
    unblockSlot: async (slotId: string): Promise<SlotResponse> => {
        const response = await apiClient.patch<SlotResponse>(`/slots/${slotId}/unblock`)
        return response.data
    },
}
